### ISSUE TITLE
Can't use Object.entries with ...- rest destructuring

### ISSUE THREAD

**Issue by safareli** [microsoft/TypeScript#46385]
# Bug Report

### 🔎 Search Terms
- Object.entries, Object.values, destructuring
 

### 🕗 Version & Regression Information

- This is a **type-system error**
- This is the behavior in every version I tried **[ 3.3.3 - 4.5beta ]**
- I've reviewed the FAQ for entries about **destructuring**

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.3#code/C4TwDgpgBAIghsOUC8UDeATBcBcUDOwATgJYB2A5gL4BQokUAGiujVO1AGYkBeA-HniIoAHygBXMhgjcyEDAG42HAEZx+g7KIlSZ5eUo5QVvAbC1jJ02QeXsVAew3nhl3TcU1aNaQGMANnBE0JySvsAkDmTs+CQAtmD+EADyKgBWEOEAomTEJBD4ABRBFDhoUADa+HiEpJQAupquOtb6GFBUAJQ4AG4OJBg+mYHBUL5RhFAAHniMSuNkk2gAdKtTAExULFNKNLEJSakZ2bmkBYVTnUr7iSnpmcA5eecbVzQA9O9G3z8cAH4Aj5fACCRAo4jiEFyUAcnCg9GgAHJytxnEIkG5WnJFMZ1GZ0dorHpsQpjKYmhiWsSDMYnPiLFSPKSqIioCR8FAyA5gFA4PhYhQyHAVEl4Q4oGAgnBIcAIEQYXCEVBkZVqgQ8g0KczEcsgRwAJK6KYEEiChDiUbsqBxdmxShs6JKlWo+nNIlM3Fohnutqkkxet3uX20gOUn0kjo6wrrADMABYAKydGhAA)

### 💻 Code

```ts
type Data = {data: string}
type X = {
    fiz?: Data | undefined;
    baz?: Data | undefined;
    biz?: Data | undefined;
    boz?: Data | undefined;
}

declare function  simpleObjectEntries(arg:{ [s: string]: Data | undefined }):void
declare const x: X;
const {...x2} = x;

simpleObjectEntries(x);
simpleObjectEntries(x2);
//                  ~~~
// Argument of type '{ fiz?: Data | undefined; baz?: Data | undefined; biz?: Data | undefined; boz?: Data | undefined; }' is not assignable to parameter of type '{ [s: string]: Data; }'.
//   Index signature is missing in type '{ fiz?: Data | undefined; baz?: Data | undefined; biz?: Data | undefined; boz?: Data | undefined; }'.(2345)
```

### 🙁 Actual behavior
Actual behavior is the error that shouldn't be raised. if it should be raised then the non destructed x should be as invalid as the destructed.

### 🙂 Expected behavior
it's expected to not get the type error at all. as `simpleObjectEntries(x);` is accepted.

**Comment by jcalz** [comment]
duplicate #42021 I think?

**Comment by gabritto** [comment]
I think this is a duplicate of #42021, unless I'm missing something: `x` has an implicit index signature because its type is an object type (`X`), but when you rest destructure `x` into `x2`, we consider `x2` to no longer have the implicit index signature, hence the error message. There's a proposal to fix this: #48014.
